src/rabin.js

```diff
@@ -0,0 +1,186 @@
+            result = (result * base) % modulus;
+        exponent = exponent >> 1n; //divide by 2
+        base = (base * base) % modulus;
+    }
+    return result;
+}
+
+function rabinHashBytes(bytes) {
+    hBytes = bsv.crypto.Hash.sha256(bytes);
+    let idx = hBytes.byteLength / 2;
+    let hl = bsv.crypto.Hash.sha256(hBytes.slice(0, idx));
+    let hr = bsv.crypto.Hash.sha256(hBytes.slice(idx, hBytes.byteLength));
+    return toBigIntLE(Buffer.concat([hl, hr]));
+}
+
+function calculateNextPrime(p) {
+    smallPrimesProduct = 3n * 5n * 7n * 11n * 13n * 17n * 19n * 23n * 29n;
+    while (greatestCommonDivisor(p, smallPrimesProduct) != 1) {
+        p = p + 4n;
+    }
+    if (powerMod(2n, p - 1n, p) != 1n) {
+        return calculateNextPrime(p + 4n);
+    }
+    if (powerMod(3n, p - 1n, p) != 1n) {
+        return calculateNextPrime(p + 4n);
+    }
+    if (powerMod(5n, p - 1n, p) != 1n) {
+        return calculateNextPrime(p + 4n);
+    }
+    if (powerMod(17n, p - 1n, p) != 1n) {
+        return calculateNextPrime(p + 4n);
+    }
+    return p;
+}
+
+function getPrimeNumber(p) {
+    while (p % 4n != 3n) {
+        p = p + 1n;
+    }
+    return calculateNextPrime(p);
+}
+
+const paddingBuffer = Buffer.from('00', 'hex');
+
+function root(dataBuffer, p, q, nRabin) {
+    let sig, x, paddingByteCount = 0;
+    while (true) {
+        let h1 = rabinHashBytes(dataBuffer);
+        x = rabinHashBytes(dataBuffer) % nRabin;
+        sig = powerMod(p, q - 2n, q) * p * powerMod(x, (q + 1n) / 4n, q);
+        sig = (powerMod(q, p - 2n, p) * q * powerMod(x, (p + 1n) / 4n, p) + sig) % (nRabin);
+        if (((sig * sig) % nRabin) === x) {
+            break;
+        }
+        dataBuffer = Buffer.concat([dataBuffer, paddingBuffer]);
+        paddingByteCount++;
+    }
+    return {
+        "signature": decimalToHexString(sig),
+        "paddingByteCount": paddingByteCount
+    };
+}
+
+function calculateRabinValueFromKeyParts(p,q){
+    if((typeof(p) !== 'bigint' && isNaN(p)) || (typeof(q) !== 'bigint' && isNaN(q)))
+        throw("Error: Key parts should be numbers.")
+    return decimalToHexString(BigInt(p) * BigInt(q));
+}
+
+/**
+ * Generates Key p, q and nRabin values from a hexadecimal seed
+ * @param {String} seed hexadecimal string value
+ * @returns {JSON} {'p': String,'q': String, 'nRabin': String} Key p, q and nRabin values in hex
+ */
+function generateRabinKeyFromSeed(seed) {
+    // Check if seed is valid hex
+    if (!checkIfValidHexString(seed))
+        throw ("Error: Seed %s should be a hexadecimal String with or without '0x' at the beginning.", seed);
+    // Remove 0x from seed if necessary
+    seed = seed.replace('0x', '');
+
+    let p = getPrimeNumber(rabinHashBytes(Buffer.from(seed, 'hex')) % ((2n ** 501n) + 1n));
+    let q = getPrimeNumber(rabinHashBytes(Buffer.from(seed + '00', 'hex')) % ((2n ** 501n) + 1n));
+    let nRabin = calculateRabinValueFromKeyParts(p,q);
+    return {
+        "p": decimalToHexString(p),
```

Comment:
Let's just return p, so users can decide what to do with it.

Reply:
But then how would they calculate nRabin?! We will need to output the seed value anyway when the PRNG is integrated, maybe these functions should be separated? With an additional calculatePrivateFromSeed?

Reply:
I mean return p as an integer, so does q. It's user's responsibility to save the private key (p, q), from which public key nRabin can be computed. Let's not use seed, just PRNG as in bsv library.